cl/sitemaps_infinite/base_sitemap.py

```diff
@@ -0,0 +1,263 @@
+from functools import cached_property
+from typing import Any, Tuple, Union
+
+from cursor_pagination import CursorPaginator
+from django.conf import settings
+from django.contrib import sitemaps
+from django.contrib.sites.models import Site
+from django.contrib.sites.requests import RequestSite
+from django.core.paginator import InvalidPage
+from redis import Redis
+
+from cl.lib.redis_utils import get_redis_interface
+from cl.sitemaps_infinite.types import SitemapsPagesNumber
+
+logger = logging.getLogger(__name__)
+
+REDIS_DB = "CACHE"
+HASH_NAME = "large-sitemaps:num_pages"
+
+
+class CacheableList(list):
+    """
+    A wrapper around a list that can hold the additional attributes,
+    used to store the cache metadata - e.g. the `expiration_time` and `cursor` used to get that list.
+
+    Used to keep the compatibility with the old cached sitemap generation
+    and to avoid the sync problems, if cache metadata is handled in a separate cache key.
+
+    There is no way to get the cache item expiration time using Django cache API.
+    """
+
+    expiration_time: datetime
+    current_cursor: str
+    next_cursor: str
+
+
+class CustomCursorPaginator(CursorPaginator):
+    """
+    A custom CursorPaginator that uses the number of pages from the cache.
+    """
+
+    section: str | None = None
+
+    def __init__(self, queryset, ordering: dict[str], section: str):
+        super().__init__(queryset, ordering)
+        self.section = section
+
+    @cached_property
+    def num_pages(self) -> int:
+        """
+        Returns the total number of pages in the paginator, reads it from the cache.
+        Overloaded version to be compatible with the standard Django Paginator API.
+        """
+        redis_db: Redis = get_redis_interface(REDIS_DB)
+        num_pages_cached: SitemapsPagesNumber = redis_db.hgetall(HASH_NAME)
+
+        if self.section is None:
+            raise Exception("The section is not set in the paginator.")
+
+        return int(num_pages_cached.get(self.section, 1))
+
+    def save_num_pages(self, num: int) -> None:
+        """
+        Saves the number of pages for the current section to the cache.
+        """
+        redis_db: Redis = get_redis_interface(REDIS_DB)
+        redis_db.hset(HASH_NAME, self.section, num)
+        logger.debug(f"Saved {num} pages for {self.section} in the cache")
+
+
+class InfinitePaginatorSitemap(sitemaps.Sitemap):
+    """
+    A base class for the sitemap that uses seek pagination from CursorPaginator to generate a large number of URLs.
+
+    This sitemap is intended to be pre-generated by a Command rather than generated on-the-fly,
+    as accessing it directly will raise an `InvalidPage` exception.
+
+    The `get_urls_by_cursor` method is used to generate the URLs,
+    taking an optional cursor parameter to fetch the next set of URLs.
+    """
+
+    # set url scheme to be compatible with the regular sitemap
+    protocol = settings.SITEMAPS_URL_SCHEME
```

Comment:
I've looked at the [get_protocol](https://github.com/django/django/blob/0d92428d77fafff373e05dd5a6cdb62bd1dfbda0/django/contrib/sitemaps/__init__.py#L81C1-L83C52) method in the Sitemap class, and I believe we can simplify things. We don't need a new environment variable for the URL scheme. Instead, we can override the method in this class and leverage existing settings to provide the default. Here's a suggestion:

```suggestion
    def get_protocol(self, protocol=None):
        env_based_protocol = (
            "http"
            if any([settings.TESTING, settings.DEVELOPMENT])
            else "https"
        )
        return self.protocol or protocol or env_based_protocol
```